### Moive Recommendation
In this notebook, we will use an Alternating Least Squares (ALS) algorithm with Spark APIs to predict the ratings for the movies.

In [2]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import math
%matplotlib inline

In [3]:
import os
os.environ["PYSPARK_PYTHON"] = "python3"

## Part1: Data ETL and Data Exploration

In [5]:
from pyspark.sql import SparkSession
spark = SparkSession \
    .builder \
    .appName("moive analysis") \
    .config("spark.some.config.option", "some-value") \
    .getOrCreate()

In [6]:
movies_df = spark.read.load("/FileStore/tables/movies.csv", format='csv', header = True)
ratings_df = spark.read.load("/FileStore/tables/ratings.csv", format='csv', header = True)
links_df = spark.read.load("/FileStore/tables/links.csv", format='csv', header = True)
tags_df = spark.read.load("/FileStore/tables/tags.csv", format='csv', header = True)

In [7]:
movies_df.show(5)

+-------+--------------------+--------------------+
movieId| title| genres|
+-------+--------------------+--------------------+
 1| Toy Story (1995)|Adventure|Animati...|
 2| Jumanji (1995)|Adventure|Childre...|
 3|Grumpier Old Men ...| Comedy|Romance|
 4|Waiting to Exhale...|Comedy|Drama|Romance|
 5|Father of the Bri...| Comedy|
+-------+--------------------+--------------------+
only showing top 5 rows

In [8]:
ratings_df.show(100)

+------+-------+------+---------+
userId|movieId|rating|timestamp|
+------+-------+------+---------+
 1| 1| 4.0|964982703|
 1| 3| 4.0|964981247|
 1| 6| 4.0|964982224|
 1| 47| 5.0|964983815|
 1| 50| 5.0|964982931|
 1| 70| 3.0|964982400|
 1| 101| 5.0|964980868|
 1| 110| 4.0|964982176|
 1| 151| 5.0|964984041|
 1| 157| 5.0|964984100|
 1| 163| 5.0|964983650|
 1| 216| 5.0|964981208|
 1| 223| 3.0|964980985|
 1| 231| 5.0|964981179|
 1| 235| 4.0|964980908|
 1| 260| 5.0|964981680|
 1| 296| 3.0|964982967|
 1| 316| 3.0|964982310|
 1| 333| 5.0|964981179|
 1| 349| 4.0|964982563|
 1| 356| 4.0|964980962|
 1| 362| 5.0|964982588|
 1| 367| 4.0|964981710|
 1| 423| 3.0|964982363|
 1| 441| 4.0|964980868|
 1| 457| 5.0|964981909|
 1| 480| 4.0|964982346|
 1| 500| 3.0|964981208|
 1| 527| 5.0|964984002|
 1| 543| 4.0|964981179|
 1| 552| 4.0|964982653|
 1| 553| 5.0|964984153|
 1| 590| 4.0|964982546|
 1| 592| 4.0|964982271|
 1| 593| 4.0|964983793|
 1| 596| 5.0|964982838|
 1| 608| 5.0|964982931|
 1| 648| 3.0|964982563|
 1| 661| 5.0|964982838|
 1| 673| 3.0|964981775|
 1| 733| 4.0|964982400|
 1| 736| 3.0|964982653|
 1| 780| 3.0|964984086|
 1| 804| 4.0|964980499|
 1| 919| 5.0|964982475|
 1| 923| 5.0|964981529|
 1| 940| 5.0|964982176|
 1| 943| 4.0|964983614|
 1| 954| 5.0|964983219|
 1| 1009| 3.0|964981775|
 1| 1023| 5.0|964982681|
 1| 1024| 5.0|964982876|
 1| 1025| 5.0|964982791|
 1| 1029| 5.0|964982855|
 1| 1030| 3.0|964982903|
 1| 1031| 5.0|964982653|
 1| 1032| 5.0|964982791|
 1| 1042| 4.0|964981179|
 1| 1049| 5.0|964982400|
 1| 1060| 4.0|964980924|
 1| 1073| 5.0|964981680|
 1| 1080| 5.0|964981327|
 1| 1089| 5.0|964982951|
 1| 1090| 4.0|964984018|
 1| 1092| 5.0|964983484|
 1| 1097| 5.0|964981680|
 1| 1127| 4.0|964982513|
 1| 1136| 5.0|964981327|
 1| 1196| 5.0|964981827|
 1| 1197| 5.0|964981872|
 1| 1198| 5.0|964981827|
 1| 1206| 5.0|964983737|
 1| 1208| 4.0|964983250|
 1| 1210| 5.0|964980499|
 1| 1213| 5.0|964982951|
 1| 1214| 4.0|964981855|
 1| 1219| 2.0|964983393|
 1| 1220| 5.0|964981909|
 1| 1222| 5.0|964981909|
 1| 1224| 5.0|964984018|
 1| 1226| 5.0|964983618|
 1| 1240| 5.0|964983723|
 1| 1256| 5.0|964981442|
 1| 1258| 3.0|964983414|
 1| 1265| 4.0|964983599|
 1| 1270| 5.0|964983705|
 1| 1275| 5.0|964982290|
 1| 1278| 5.0|964983414|
 1| 1282| 5.0|964982703|
 1| 1291| 5.0|964981909|
 1| 1298| 5.0|964984086|
 1| 1348| 4.0|964983393|
 1| 1377| 3.0|964982653|
 1| 1396| 3.0|964983017|
 1| 1408| 3.0|964982310|
 1| 1445| 3.0|964984112|
 1| 1473| 4.0|964980875|
 1| 1500| 4.0|964980985|
 1| 1517| 5.0|964981107|
 1| 1552| 4.0|964982620|
+------+-------+------+---------+
only showing top 100 rows

In [9]:
links_df.show(5)

+-------+-------+------+
movieId| imdbId|tmdbId|
+-------+-------+------+
 1|0114709| 862|
 2|0113497| 8844|
 3|0113228| 15602|
 4|0114885| 31357|
 5|0113041| 11862|
+-------+-------+------+
only showing top 5 rows

In [10]:
tags_df.show(5)

+------+-------+---------------+----------+
userId|movieId| tag| timestamp|
+------+-------+---------------+----------+
 2| 60756| funny|1445714994|
 2| 60756|Highly quotable|1445714996|
 2| 60756| will ferrell|1445714992|
 2| 89774| Boxing story|1445715207|
 2| 89774| MMA|1445715200|
+------+-------+---------------+----------+
only showing top 5 rows

In [11]:
tmp1 = ratings_df.groupBy("userID").count().toPandas()['count'].min()
tmp2 = ratings_df.groupBy("movieId").count().toPandas()['count'].min()
print('For the users that rated movies and the movies that were rated:')
print('Minimum number of ratings per user is {}'.format(tmp1))
print('Minimum number of ratings per movie is {}'.format(tmp2))

For the users that rated movies and the movies that were rated:
Minimum number of ratings per user is 20
Minimum number of ratings per movie is 1

In [12]:
tmp1 = sum(ratings_df.groupBy("movieId").count().toPandas()['count'] == 1)
tmp2 = ratings_df.select('movieId').distinct().count()
print('{} out of {} movies are rated by only one user'.format(tmp1, tmp2))

3446 out of 9724 movies are rated by only one user

## Part 1: Spark SQL and OLAP

In [14]:
movies_df.registerTempTable("movies")
ratings_df.registerTempTable("ratings")
links_df.registerTempTable("links")
tags_df.registerTempTable("tags")


### Q1: The number of Users

In [16]:
%sql 
Select Count(Distinct userId) as Number_of_users From ratings

Number_of_users 610

### Q2: The number of Movies

In [18]:
%sql 
Select Count(Distinct movieId) as Number_of_movies From movies

Number_of_movies 9742

### Q3:  How many movies are rated by users? List movies not rated before

In [20]:
%sql
 
Select title, genres from movies  
where movieId not in 
  (Select movieId from ratings)



title genres Innocents, The (1961) Drama|Horror|Thriller Niagara (1953) Drama|Thriller For All Mankind (1989) Documentary Color of Paradise, The (Rang-e khoda) (1999) Drama I Know Where I'm Going! (1945) Drama|Romance|War Chosen, The (1981) Drama Road Home, The (Wo de fu qin mu qin) (1999) Drama|Romance Scrooge (1970) Drama|Fantasy|Musical Proof (1991) Comedy|Drama|Romance Parallax View, The (1974) Thriller This Gun for Hire (1942) Crime|Film-Noir|Thriller Roaring Twenties, The (1939) Crime|Drama|Thriller Mutiny on the Bounty (1962) Adventure|Drama|Romance In the Realms of the Unreal (2004) Animation|Documentary Twentieth Century (1934) Comedy Call Northside 777 (1948) Crime|Drama|Film-Noir Browning Version, The (1951) Drama Chalet Girl (2011) Comedy|Romance

In [21]:
%sql 
Select Count(Distinct movieID) as Number_movies_are_rated_by_users From ratings

Number_movies_are_rated_by_users 9724

### Q4: List Movie Genres

In [23]:
%sql
Select Distinct genres From movies

genres Comedy|Horror|Thriller Adventure|Sci-Fi|Thriller Action|Adventure|Drama|Fantasy Action|Drama|Horror Action|Animation|Comedy|Sci-Fi Animation|Children|Drama|Musical|Romance Action|Adventure|Drama Adventure|Sci-Fi Documentary|Musical|IMAX Adventure|Children|Fantasy|Sci-Fi|Thriller Adventure|Animation Musical|Romance|War Action|Adventure|Fantasy|Romance Adventure|Children|Drama|Fantasy|IMAX Comedy|Crime|Horror|Thriller Crime|Drama|Fantasy|Horror|Thriller Comedy|Mystery|Thriller Adventure|Fantasy Horror|Romance|Sci-Fi Drama|Film-Noir|Romance Crime|Drama|Film-Noir|Mystery|Thriller Musical|Western Adventure|Animation|Children|Musical|Romance Comedy|Crime|Mystery|Romance|Thriller Action|Animation|Children|Crime Adventure|Horror|Sci-Fi Comedy|Drama|Romance|Thriller Action|Drama|Mystery|Sci-Fi|Thriller Adventure|Animation|Children|Comedy|Drama|Fantasy Action|Adventure|Mystery|Sci-Fi Adventure|Drama|Romance|Thriller|War Animation|Drama Horror|Thriller|IMAX Crime Action|Animation|Film-Noir|Sci-Fi|Thriller Action|Crime|Horror|Sci-Fi|Thriller Crime|Drama|Thriller Action|Adventure|Animation|Sci-Fi Action|Adventure|Drama|Fantasy|Mystery|IMAX Mystery|Sci-Fi|Thriller Adventure|Drama|IMAX Romance Comedy|Horror|Musical|Sci-Fi Children|Drama|Musical Action|Drama Crime|Thriller Drama|Mystery|Sci-Fi|Thriller Comedy|Crime|Drama|Western Action|Adventure|Drama|Fantasy|Romance|Sci-Fi|Thriller Action|Fantasy|Horror|IMAX Action|Sci-Fi Adventure|Crime|Drama|Romance Drama|Fantasy|Mystery|Sci-Fi Action|Crime|Mystery Animation|Children|Comedy|Fantasy|Musical|Romance|IMAX Action|Adventure|Sci-Fi|War|IMAX Comedy|Drama|Fantasy|Mystery|Romance Children|Comedy|Crime|Drama|Fantasy Action|Comedy|Fantasy|IMAX Action|Adventure|Fantasy Thriller Comedy|Fantasy|Romance Romance|Thriller Adventure|Children|Comedy|Mystery Drama|Fantasy|Romance|Sci-Fi Documentary|Drama|Thriller Crime|Drama|Horror|Mystery|Thriller Action|Romance|Thriller Action|Crime|Drama|Sci-Fi|Thriller Film-Noir|Mystery|Thriller Crime|Thriller|War Comedy|Drama|Fantasy|Romance Adventure|Animation|Children|Comedy|Fantasy|Romance Comedy|Crime|Drama|War Adventure|Animation|Fantasy|Romance Drama|Fantasy|Horror|Romance|Thriller Action|Sci-Fi|Thriller|Western|IMAX Animation|Children Action|Crime|Mystery|Thriller Comedy|Romance|Sci-Fi Adventure|War|Western Action|Animation|Comedy|Fantasy Comedy|Horror|Romance Comedy|Drama|Musical Drama|Film-Noir|Mystery|Romance Drama|Musical|Romance Adventure|Comedy|Thriller Action|Animation|Children|Comedy|Sci-Fi|IMAX Drama|Sci-Fi|IMAX Animation|Children|Mystery Action|Sci-Fi|Thriller|Western Action|Adventure|Animation|Comedy|Crime|Mystery Action|Adventure|Animation|Comedy|Fantasy|Mystery|Sci-Fi Adventure|Comedy|Drama|Romance Action|Adventure|Animation|Children|Comedy|Western Comedy|Horror Animation|Children|Comedy|Romance Animation|Children|Comedy|Musical|Romance Adventure|Animation|Fantasy Children|Comedy|Crime|Musical Action|Adventure|Drama|Sci-Fi|Thriller Action|Comedy|Crime|Mystery Crime|Horror|Thriller Adventure|Comedy|Sci-Fi Animation|Children|Comedy|Fantasy|Musical|Romance Action|Adventure|Drama|Sci-Fi Action|Drama|Horror|Thriller Adventure|Children Animation|Fantasy|Thriller Comedy|Drama|Horror|Sci-Fi|Thriller Action|Thriller|War Action|Crime|Mystery|Sci-Fi|Thriller Adventure|Animation|Children|Musical|Western Crime|Drama|Fantasy|Mystery|Thriller Animation|Comedy|Horror|IMAX Action|Comedy|Crime|Fantasy Adventure|Animation|Comedy|Crime Action|Adventure|Fantasy|Thriller Action|Adventure|Fantasy|Horror Action|Animation|Sci-Fi Action|Adventure|Drama|Romance|Thriller|Western Adventure|Comedy|Fantasy|Musical Action|Adventure|Animation|Children|Comedy|Sci-Fi Action|Adventure|Crime|IMAX Action|Crime|Thriller Crime|Drama Adventure|Comedy|Western Documentary|Drama Romance|Sci-Fi Adventure|Drama|Horror|Sci-Fi|Thriller Drama|Sci-Fi|War Action|Comedy|Drama|Horror|Thriller Action|Comedy|Crime|Horror Comedy|Mystery|Romance Comedy|Musical|Romance Fantasy|Sci-Fi Action|Animation|

In [24]:
%sql
SELECT distinct explode(split(genres,'[|]')) as category
FROM movies 


category Crime Romance Thriller Adventure Drama War Documentary Fantasy Mystery Musical Animation Film-Noir (no genres listed) IMAX Horror Western Comedy Children Action Sci-Fi

### Q5: Movie for Each Category

In [26]:
%sql
Select category, count(distinct movieID) as number
From
(
SELECT movieID, title, explode(split(genres,'[|]')) as category
FROM movies
) as cte
Group by category
order by count(distinct movieID)



category number (no genres listed) 34 Film-Noir 87 IMAX 158 Western 167 Musical 334 War 382 Documentary 440 Mystery 573 Animation 611 Children 664 Fantasy 779 Horror 978 Sci-Fi 980 Crime 1199 Adventure 1263 Romance 1596 Action 1828 Thriller 1894 Comedy 3756 Drama 4361

In [27]:
%sql 

Select category, collect_list(distinct title) as movies
From
(
SELECT movieID, title, explode(split(genres,'[|]')) as category
FROM movies
) as cte
Group by category
order by category

category movies (no genres listed) List(Guardians (2016), Paterson, Hyena Road, A Cosmic Christmas (1977), Too Funny to Fail: The Life and Death of The Dana Carvey Show (2017), A Christmas Story Live! (2017), The Adventures of Sherlock Holmes and Doctor Watson, A Midsummer Night's Dream (2016), Death Note: Desu nôto (2006–2007), Pirates of the Caribbean: Dead Men Tell No Tales (2017), Whiplash (2013), The Putin Interviews (2017), T2 3-D: Battle Across Time (1996), Green Room (2015), Grease Live (2016), La cravate (1957), Maria Bamford: Old Baby, The Godfather Trilogy: 1972-1990 (1992), Cosmos, Let It Be Me (1995), Lemonade (2016), Ethel & Ernest (2016), The Brand New Testament (2015), Noin 7 veljestä (1968), The Adventures of Sherlock Holmes and Doctor Watson: The Hunt for the Tiger (1980), Superfast! (2015), Ali Wong: Baby Cobra (2016), The Forbidden Dance (1990), Serving in Silence: The Margarethe Cammermeyer Story (1995), Trevor Noah: African American (2013), Ben-hur (2016), Generation Iron 2, The OA, Black Mirror) Action List(Superman III (1983), Impostor (2002), Die Another Day (2002), Italian Job, The (2003), RocknRolla (2008), Daybreakers (2010), Book of Eli, The (2010), Snow White and the Huntsman (2012), Captain America: Civil War (2016), Iron Man & Hulk: Heroes United (2013), Ryuzo and the Seven Henchmen (2015), Ride Along 2 (2016), Star Wars: Episode III - Revenge of the Sith (2005), Day Watch (Dnevnoy dozor) (2006), Hell Ride (2008), A-Team, The (2010), Hot Pursuit (2015), Bright (2017), Men in Black (a.k.a. MIB) (1997), Dragonslayer (1981), Clash of the Titans (1981), Taxi (2004), Wraith, The (1986), Mortadelo & Filemon: The Big Adventure (La gran aventura de Mortadelo y Filemón) (2003), Homefront (2013), London Has Fallen (2016), Dredd (2012), LEGO Batman: The Movie - DC Heroes Unite (2013), Gen-X Cops (1999), Treasure of the Sierra Madre, The (1948), Assignment, The (1997), eXistenZ (1999), Sanjuro (Tsubaki Sanjûrô) (1962), Patriot Games (1992), Another 48 Hrs. (1990), Ice Harvest, The (2005), 300 (2007), MacGruber (2010), Ghost in the Shell: Stand Alone Complex - The Laughing Man (2005), Bad Asses on the Bayou (2015), Judge Dredd (1995), Mad Max (1979), Casino Royale (1967), Bandidas (2006), Dead Fury (2008), Last Airbender, The (2010), Evangelion: 3.0 You Can (Not) Redo (2012), Band of Brothers (2001), Natural Born Killers (1994), Battle Creek Brawl (Big Brawl, The) (1980), Four Musketeers, The (1974), Crank (2006), Ghost Rider (2007), Tekkonkinkreet (Tekkon kinkurîto) (2006), Terminator Salvation (2009), Gravity (2013), Maze Runner, The (2014), The Last Witch Hunter (2015), Freddy vs. Jason (2003), Where Eagles Dare (1968), Paycheck (2003), Fay Grim (2006), Watchmen (2009), Justice League: Doom (2012) , Escape Plan (2013), The Flash 2 - Revenge of the Trickster (1991), Maximum Ride (2016), Love and a .45 (1994), King Kong vs. Godzilla (Kingukongu tai Gojira) (1962), Moonraker (1979), Big Trouble in Little China (1986), Extreme Days (2001), Godzilla vs. Mothra (Mosura tai Gojira) (1964), Public Enemy, The (1931), Superman/Batman: Public Enemies (2009), Fullmetal Alchemist: The Sacred Star of Milos (2011), Midnight Chronicles (2009), Stretch (2014), Mortal Kombat: The Journey Begins (1995), Punisher, The (1989), Kangaroo Jack (2003), Good, the Bad and the Ugly, The (Buono, il brutto, il cattivo, Il) (1966), Boot, Das (Boat, The) (1981), Blade (1998), For a Few Dollars More (Per qualche dollaro in più) (1965), Kiss of the Dragon (2001), Terminator 3: Rise of the Machines (2003), Touch of Zen, A (Xia nu) (1971), Rango (2011), Lockout (2012), Good Day to Die Hard, A (2013), Need for Speed (2014), In the Name of the King III (2014), Army of Darkness (1993), Speed 2: Cruise Control (1997), Armour of God II: Operation Condor (Operation Condor) (Fei ying gai wak) (1991), Innerspace (1987), Troy (2004), Team America: World Police (2004), Hitman (2007), Echelon Conspiracy (2009), Repo Men (2010), The Hitman's Bodyguard 

## Part2: Spark ALS based approach for training model
We will use an Spark ML to predict the ratings, so let's reload "ratings.csv" using ``sc.textFile`` and then convert it to the form of (user, item, rating) tuples.

In [29]:
ratings_df.show()

+------+-------+------+---------+
userId|movieId|rating|timestamp|
+------+-------+------+---------+
 1| 1| 4.0|964982703|
 1| 3| 4.0|964981247|
 1| 6| 4.0|964982224|
 1| 47| 5.0|964983815|
 1| 50| 5.0|964982931|
 1| 70| 3.0|964982400|
 1| 101| 5.0|964980868|
 1| 110| 4.0|964982176|
 1| 151| 5.0|964984041|
 1| 157| 5.0|964984100|
 1| 163| 5.0|964983650|
 1| 216| 5.0|964981208|
 1| 223| 3.0|964980985|
 1| 231| 5.0|964981179|
 1| 235| 4.0|964980908|
 1| 260| 5.0|964981680|
 1| 296| 3.0|964982967|
 1| 316| 3.0|964982310|
 1| 333| 5.0|964981179|
 1| 349| 4.0|964982563|
+------+-------+------+---------+
only showing top 20 rows

In [30]:
movie_ratings=ratings_df.drop('timestamp')

In [31]:
# Data type convert
from pyspark.sql.types import IntegerType, FloatType
movie_ratings = movie_ratings.withColumn("userId", movie_ratings["userId"].cast(IntegerType()))
movie_ratings = movie_ratings.withColumn("movieId", movie_ratings["movieId"].cast(IntegerType()))
movie_ratings = movie_ratings.withColumn("rating", movie_ratings["rating"].cast(FloatType()))

In [32]:
movie_ratings.show()

+------+-------+------+
userId|movieId|rating|
+------+-------+------+
 1| 1| 4.0|
 1| 3| 4.0|
 1| 6| 4.0|
 1| 47| 5.0|
 1| 50| 5.0|
 1| 70| 3.0|
 1| 101| 5.0|
 1| 110| 4.0|
 1| 151| 5.0|
 1| 157| 5.0|
 1| 163| 5.0|
 1| 216| 5.0|
 1| 223| 3.0|
 1| 231| 5.0|
 1| 235| 4.0|
 1| 260| 5.0|
 1| 296| 3.0|
 1| 316| 3.0|
 1| 333| 5.0|
 1| 349| 4.0|
+------+-------+------+
only showing top 20 rows

### ALS Model Selection and Evaluation

With the ALS model, we can use a grid search to find the optimal hyperparameters.

In [34]:
# import package
from pyspark.ml.evaluation import RegressionEvaluator
from pyspark.ml.recommendation import ALS
from pyspark.ml.tuning import CrossValidator,ParamGridBuilder

In [35]:
#Create test and train set
(training,test)=movie_ratings.randomSplit([0.8,0.2])

In [36]:
#Create ALS model
als = ALS(userCol="userId", itemCol="movieId", ratingCol ="rating", coldStartStrategy = "drop", nonnegative = True)

In [37]:
#Tune model using ParamGridBuilder
param_grid = ParamGridBuilder().addGrid(als.rank, [10,11,12]).addGrid(als.maxIter, [5,7,10]).addGrid(als.regParam, [0.1, 0.01, 0.001]).build()

In [38]:
param_grid = ParamGridBuilder().addGrid(als.rank, [10,11,12]).addGrid(als.maxIter, [5,7,10]).addGrid(als.regParam, [0.1, 0.01, 0.001]).build()
evaluator = RegressionEvaluator(metricName= "rmse", labelCol="rating", predictionCol="prediction")
crossval = CrossValidator(estimator=als,
                          estimatorParamMaps=param_grid,
                          evaluator=evaluator,
                          numFolds=5)

In [39]:
# Define evaluator as RMSE
evaluator = RegressionEvaluator(metricName= "rmse", labelCol="rating", predictionCol="prediction")


In [40]:
# Build Cross validation 
crossval = CrossValidator(estimator=als,
                          estimatorParamMaps=param_grid,
                          evaluator=evaluator,
                          numFolds=5)


In [41]:
#Fit ALS model to training data
model = crossval.fit(training)

/databricks/spark/python/pyspark/ml/util.py:791: UserWarning: Can not find mlflow. To enable mlflow logging, install MLflow library from PyPi.
 warnings.warn(_MLflowInstrumentation._NO_MLFLOW_WARNING)

In [42]:
model = crossval.fit(training)
best_model = model.bestModel

/databricks/spark/python/pyspark/ml/util.py:791: UserWarning: Can not find mlflow. To enable mlflow logging, install MLflow library from PyPi.
 warnings.warn(_MLflowInstrumentation._NO_MLFLOW_WARNING)

In [43]:
#Extract best model from the tuning exercise using ParamGridBuilder
best_model = model.bestModel

### Model testing
And finally, make a prediction and check the testing error.

In [45]:
#Generate predictions and evaluate using RMSE
predictions=best_model.transform(test)
rmse = evaluator.evaluate(predictions)

In [46]:
#Print evaluation metrics and model parameters
print ("RMSE = "+str(rmse))
print ("**Best Model**")
print (" Rank:" +str(best_model.rank))
print (" MaxIter:" +str(best_model._java_obj.parent().getMaxIter()))
print (" RegParam:" + str(best_model._java_obj.parent().getRegParam()))

RMSE = 0.8696467860575481
**Best Model**
 Rank:12
 MaxIter:10
 RegParam:0.1

In [47]:
predictions.show()

+------+-------+------+----------+
userId|movieId|rating|prediction|
+------+-------+------+----------+
 599| 471| 2.5| 2.6550376|
 409| 471| 3.0| 3.7558448|
 387| 471| 3.0| 3.0120077|
 448| 471| 4.0| 3.1823394|
 541| 471| 3.0| 3.989418|
 104| 471| 4.5| 3.8197446|
 609| 833| 3.0| 1.1993757|
 492| 833| 4.0| 1.3983767|
 599| 1088| 2.5| 2.3889964|
 414| 1088| 3.0| 2.9139524|
 517| 1088| 1.0| 2.876779|
 177| 1088| 3.5| 3.8400588|
 84| 1088| 3.0| 3.3194344|
 554| 1088| 5.0| 4.1464996|
 594| 1088| 4.5| 4.5328407|
 525| 1088| 4.5| 3.3773742|
 587| 1238| 4.0| 3.094621|
 216| 1238| 5.0| 3.3964276|
 387| 1342| 3.0| 2.5852082|
 19| 1342| 2.0| 2.721101|
+------+-------+------+----------+
only showing top 20 rows

### Model apply and see the performance

In [49]:
alldata=best_model.transform(movie_ratings)
rmse = evaluator.evaluate(alldata)
print ("RMSE = "+str(rmse))

RMSE = 0.6376846849381629

In [50]:
alldata.registerTempTable("alldata")

In [51]:
%sql select * from alldata

userId movieId rating prediction 191 148 5.0 4.896819 372 471 3.0 2.9016063 603 471 4.0 3.3372178 136 471 4.0 3.9194884 469 471 5.0 3.923933 414 471 5.0 3.4931886 608 471 1.5 2.4006622 104 471 4.5 3.8197446 133 471 4.0 3.3036246 597 471 2.0 3.6391363 436 471 3.0 3.5063028 602 471 4.0 3.3714476 91 471 1.0 2.4851396 409 471 3.0 3.7558448 474 471 3.0 3.6820312 500 471 1.0 2.1650207 182 471 4.5 4.310622 385 471 4.0 3.1878355 217 471 2.0 2.9235692 599 471 2.5 2.6550376 176 471 5.0 3.9440436 520 471 5.0 3.902517 462 471 2.5 2.4522321 218 471 4.0 3.4282734 541 471 3.0 3.989418 387 471 3.0 3.0120077 373 471 5.0 4.080469 57 471 3.0 3.47903 312 471 4.0 3.7873006 171 471 3.0 3.886203 273 471 5.0 4.4347997 287 471 4.5 3.8235006 610 471 4.0 3.4178226 357 471 3.5 3.9817202 555 471 3.0 3.4365625 411 471 4.0 3.3982475 32 471 3.0 3.9183002 448 471 4.0 3.1823394 216 471 3.0 2.8851533 260 471 4.5 3.8307393 426 471 5.0 3.9022813 191 496 5.0 4.896819 609 833 3.0 1.1993757 492 833 4.0 1.3983767 608 833 0.5 0.62086004 44 833 2.0 1.8911531 599 833 1.5 1.2721723 307 833 1.0 1.0165787 463 1088 3.5 3.2966218 47 1088 4.0 2.8488402 479 1088 4.0 3.4974158 41 1088 1.5 2.5282671 381 1088 3.5 3.6540904 84 1088 3.0 3.3194344 226 1088 1.0 2.8271024 414 1088 3.0 2.9139524 116 1088 4.5 3.7945027 42 1088 3.0 2.8953218 104 1088 3.0 3.682189 483 1088 3.0 3.224708 159 1088 4.0 3.6438887 606 1088 3.0 3.161968 111 1088 3.0 3.0906396 177 1088 3.5 3.8400588 474 1088 3.5 3.1962497 169 1088 4.5 4.1079116 132 1088 4.0 2.677181 554 1088 5.0 4.1464996 599 1088 2.5 2.3889964 489 1088 4.5 3.4207437 307 1088 3.0 2.3821764 387 1088 1.5 2.2819476 600 1088 3.5 2.648678 20 1088 4.5 3.812749 51 1088 4.0 3.7771816 391 1088 1.0 2.5510054 64 1088 4.0 3.2056894 563 1088 4.0 3.4439583 517 1088 1.0 2.876779 509 1088 3.0 3.119236 221 1088 3.0 2.9099536 10 1088 3.0 3.1600091 286 1088 3.5 3.3827689 200 1088 4.0 3.7195463 188 1088 4.0 3.8951516 583 1088 3.5 3.5960698 594 1088 4.5 4.5328407 555 1088 4.0 3.9302044 68 1088 3.5 3.5529244 525 1088 4.5 3.3773742 469 1238 5.0 4.293621 587 1238 4.0 3.094621 325 1238 4.0 4.1498156 19 1238 3.0 2.7696083 268 1238 5.0 4.355639 385 1238 3.0 3.4685721 462 1238 3.5 3.6551447 216 1238 5.0 3.3964276 156 1238 4.0 4.0762095 223 1342 1.0 1.5897815 425 1342 3.5 2.920737 608 1342 2.0 2.1448412 325 1342 4.0 3.3840716 19 1342 2.0 2.721101 599 1342 2.5 2.4131985 307 1342 2.0 1.8942441 387 1342 3.0 2.5852082 600 1342 2.5 2.172115 312 1342 4.0 3.5217974 503 1342 1.0 1.8356026 155 1580 4.0 3.7629735 330 1580 4.0 3.005347 47 1580 1.5 2.3986428 603 1580 4.0 3.154638 283 1580 4.0 3.6176443 359 1580 2.0 2.9058812 292 1580 4.0 3.5739858 339 1580 2.0 3.3729818 48 1580 5.0 4.1861234 266 1580 3.0 3.0230396 542 1580 3.5 3.705258 560 1580 4.0 3.290526 154 1580 3.5 4.007711 425 1580 3.5 3.2165768 586 1580 5.0 4.067678 381 1580 4.0 3.3962474 4 1580 3.0 2.6637042 59 1580 3.0 3.4454105 453 1580 4.0 3.8396564 219 1580 3.0 3.1760993 226 1580 3.5 3.2827804 21 1580 4.5 3.5318418 352 1580 2.5 3.6903248 469 1580 3.0 2.9341385 141 1580 3.5 3.3261118 414 1580 4.0 3.3597693 608 1580 3.5 3.2060885 42 1580 4.0 3.6915479 119 1580 4.0 3.698524 573 1580 4.5 4.1187787 104 1580 3.0 3.1859949 483 1580 2.0 3.0972562 322 1580 3.5 3.343499 593 1580 1.5 2.6395993 597 1580 3.0 3.60607 34 1580 2.5 2.8676891 368 1580 3.0 3.0584419 115 1580 4.0 3.6441135 28 1580 3.0 2.8169844 183 1580 4.0 3.7729855 27 1580 3.0 3.1880238 577 1580 3.0 3.3562365 606 1580 2.5 2.9991212 587 1580 4.0 4.076999 111 1580 3.0 2.5019906 230 1580 3.5 2.9337842 137 1580 3.5 3.3558898 580 1580 4.0 3.3899088 91 1580 3.5 3.3595245 140 1580 3.0 3.487874 177 1580 3.5 3.369645 416 1580 2.0 1.7737174 474 1580 4.5 3.1711543 169 1580 3.5 4.112398 500 1580 4.0 3.508209 93 1580 5.0 4.511077 232 1580 3.5 3.422223 233 1580 3.0 2.7501485 452 1580 4.0 4.4355283 274 1580 3.0 3.3195095 559 1580 5.0 3.2759292 19 1580 2.0 2.79544 182 1580 2.5 2.9045935 268 1580 3.0 2.5206692 217 1580 5.0 3.1116598 367 1580 4.0 3.6553578 599 1580 3.0 2.7040153 415 1580 3.5 3

In [52]:
%sql select * from movies join alldata on movies.movieId=alldata.movieId

movieId title genres userId movieId rating prediction 148 Awfully Big Adventure, An (1995) Drama 191 148 5.0 4.896819 471 Hudsucker Proxy, The (1994) Comedy 91 471 1.0 2.4851396 471 Hudsucker Proxy, The (1994) Comedy 409 471 3.0 3.7558448 471 Hudsucker Proxy, The (1994) Comedy 182 471 4.5 4.310622 471 Hudsucker Proxy, The (1994) Comedy 462 471 2.5 2.4522321 471 Hudsucker Proxy, The (1994) Comedy 217 471 2.0 2.9235692 471 Hudsucker Proxy, The (1994) Comedy 216 471 3.0 2.8851533 471 Hudsucker Proxy, The (1994) Comedy 469 471 5.0 3.923933 471 Hudsucker Proxy, The (1994) Comedy 373 471 5.0 4.080469 471 Hudsucker Proxy, The (1994) Comedy 357 471 3.5 3.9817202 471 Hudsucker Proxy, The (1994) Comedy 597 471 2.0 3.6391363 471 Hudsucker Proxy, The (1994) Comedy 133 471 4.0 3.3036246 471 Hudsucker Proxy, The (1994) Comedy 385 471 4.0 3.1878355 471 Hudsucker Proxy, The (1994) Comedy 436 471 3.0 3.5063028 471 Hudsucker Proxy, The (1994) Comedy 599 471 2.5 2.6550376 471 Hudsucker Proxy, The (1994) Comedy 372 471 3.0 2.9016063 471 Hudsucker Proxy, The (1994) Comedy 602 471 4.0 3.3714476 471 Hudsucker Proxy, The (1994) Comedy 603 471 4.0 3.3372178 471 Hudsucker Proxy, The (1994) Comedy 474 471 3.0 3.6820312 471 Hudsucker Proxy, The (1994) Comedy 218 471 4.0 3.4282734 471 Hudsucker Proxy, The (1994) Comedy 500 471 1.0 2.1650207 471 Hudsucker Proxy, The (1994) Comedy 57 471 3.0 3.47903 471 Hudsucker Proxy, The (1994) Comedy 387 471 3.0 3.0120077 471 Hudsucker Proxy, The (1994) Comedy 610 471 4.0 3.4178226 471 Hudsucker Proxy, The (1994) Comedy 171 471 3.0 3.886203 471 Hudsucker Proxy, The (1994) Comedy 273 471 5.0 4.4347997 471 Hudsucker Proxy, The (1994) Comedy 555 471 3.0 3.4365625 471 Hudsucker Proxy, The (1994) Comedy 176 471 5.0 3.9440436 471 Hudsucker Proxy, The (1994) Comedy 520 471 5.0 3.902517 471 Hudsucker Proxy, The (1994) Comedy 136 471 4.0 3.9194884 471 Hudsucker Proxy, The (1994) Comedy 448 471 4.0 3.1823394 471 Hudsucker Proxy, The (1994) Comedy 414 471 5.0 3.4931886 471 Hudsucker Proxy, The (1994) Comedy 608 471 1.5 2.4006622 471 Hudsucker Proxy, The (1994) Comedy 541 471 3.0 3.989418 471 Hudsucker Proxy, The (1994) Comedy 312 471 4.0 3.7873006 471 Hudsucker Proxy, The (1994) Comedy 287 471 4.5 3.8235006 471 Hudsucker Proxy, The (1994) Comedy 411 471 4.0 3.3982475 471 Hudsucker Proxy, The (1994) Comedy 426 471 5.0 3.9022813 471 Hudsucker Proxy, The (1994) Comedy 260 471 4.5 3.8307393 471 Hudsucker Proxy, The (1994) Comedy 32 471 3.0 3.9183002 471 Hudsucker Proxy, The (1994) Comedy 104 471 4.5 3.8197446 496 What Happened Was... (1994) Comedy|Drama|Romance|Thriller 191 496 5.0 4.896819 833 High School High (1996) Comedy 44 833 2.0 1.8911531 833 High School High (1996) Comedy 599 833 1.5 1.2721723 833 High School High (1996) Comedy 609 833 3.0 1.1993757 833 High School High (1996) Comedy 307 833 1.0 1.0165787 833 High School High (1996) Comedy 608 833 0.5 0.62086004 833 High School High (1996) Comedy 492 833 4.0 1.3983767 1088 Dirty Dancing (1987) Drama|Musical|Romance 132 1088 4.0 2.677181 1088 Dirty Dancing (1987) Drama|Musical|Romance 381 1088 3.5 3.6540904 1088 Dirty Dancing (1987) Drama|Musical|Romance 226 1088 1.0 2.8271024 1088 Dirty Dancing (1987) Drama|Musical|Romance 221 1088 3.0 2.9099536 1088 Dirty Dancing (1987) Drama|Musical|Romance 463 1088 3.5 3.2966218 1088 Dirty Dancing (1987) Drama|Musical|Romance 159 1088 4.0 3.6438887 1088 Dirty Dancing (1987) Drama|Musical|Romance 606 1088 3.0 3.161968 1088 Dirty Dancing (1987) Drama|Musical|Romance 599 1088 2.5 2.3889964 1088 Dirty Dancing (1987) Drama|Musical|Romance 47 1088 4.0 2.8488402 1088 Dirty Dancing (1987) Drama|Musical|Romance 111 1088 3.0 3.0906396 1088 Dirty Dancing (1987) Drama|Musical|Romance 177 1088 3.5 3.8400588 1088 Dirty Dancing (1987) Drama|Musical|Romance 474 1088 3.5 3.1962497 1088 Dirty Dancing (1987) Drama|Musical|Romance 479 1088 4.0 3.4974158 1088 Dirty Dancing (1987) Drama|Musical|Romance 20 1088 4.5 3.812749 1088 Dirty Dancing (1987) Drama|Musical

## Recommend moive to users with id: 575, 232. 
you can choose some users to recommend the moives

In [54]:
userRecs = best_model.recommendForAllUsers(10)
display(userRecs)


userId recommendations 463 List(List(3200, 4.911063), List(3022, 4.7410994), List(171495, 4.691776), List(184245, 4.691776), List(26073, 4.691776), List(72171, 4.691776), List(84273, 4.691776), List(117531, 4.691776), List(7071, 4.691776), List(179135, 4.691776)) 148 List(List(87234, 4.7610297), List(98491, 4.4837065), List(160718, 4.3590226), List(3521, 4.3576064), List(84847, 4.315737), List(96821, 4.309941), List(177593, 4.301182), List(93008, 4.2998586), List(77846, 4.2998586), List(25906, 4.2998586)) 471 List(List(177593, 4.865229), List(8477, 4.6759124), List(3200, 4.6169186), List(96004, 4.480244), List(2324, 4.447288), List(158966, 4.4352326), List(92475, 4.4244566), List(7361, 4.4241776), List(3567, 4.4173255), List(104879, 4.393123)) 496 List(List(89904, 5.164974), List(213, 4.6993084), List(32582, 4.6654897), List(60766, 4.6065784), List(55276, 4.6001487), List(4429, 4.4809093), List(6639, 4.4630513), List(6201, 4.4542937), List(4495, 4.4542937), List(5747, 4.4468584)) 540 List(List(100714, 5.103637), List(69524, 5.0196033), List(171495, 4.999335), List(179135, 4.999335), List(7071, 4.999335), List(72171, 4.999335), List(26073, 4.999335), List(84273, 4.999335), List(117531, 4.999335), List(184245, 4.999335)) 243 List(List(86377, 5.4937325), List(87234, 5.353112), List(86347, 5.330023), List(945, 5.2941575), List(5075, 5.2465158), List(8477, 5.1534834), List(3200, 5.1508293), List(7008, 5.1299267), List(26073, 5.122835), List(7071, 5.122835)) 392 List(List(5490, 5.0208697), List(5915, 4.9136357), List(3494, 4.8786855), List(1250, 4.873524), List(858, 4.8532043), List(2300, 4.8392177), List(260, 4.802099), List(174055, 4.7915807), List(3508, 4.7770834), List(112552, 4.7716937)) 31 List(List(51931, 5.6546693), List(2843, 5.6156816), List(177593, 5.399124), List(1046, 5.3527255), List(179135, 5.3485436), List(84273, 5.3485436), List(26073, 5.3485436), List(117531, 5.3485436), List(7071, 5.3485436), List(72171, 5.3485436)) 516 List(List(89904, 4.8544097), List(4429, 4.809765), List(1046, 4.5961514), List(6201, 4.5700064), List(4495, 4.5700064), List(27523, 4.5651402), List(25825, 4.5564113), List(898, 4.51829), List(177593, 4.5164113), List(5915, 4.516195)) 251 List(List(96004, 5.406329), List(3022, 5.3960986), List(26171, 5.3907647), List(177593, 5.3784337), List(2131, 5.327105), List(26471, 5.2744155), List(92475, 5.2646847), List(27156, 5.249886), List(5915, 5.246266), List(1178, 5.2420692)) 451 List(List(3200, 5.3240776), List(177593, 5.2481556), List(158966, 5.131068), List(7587, 5.083313), List(2131, 5.066201), List(6460, 5.050358), List(5915, 5.012154), List(720, 5.0090137), List(92475, 5.0028143), List(4389, 4.9831786)) 580 List(List(86347, 5.1195893), List(5075, 4.839818), List(945, 4.726414), List(3200, 4.652407), List(53123, 4.6323347), List(3837, 4.62352), List(1755, 4.620867), List(4846, 4.6095533), List(84273, 4.5717983), List(26073, 4.5717983)) 137 List(List(5915, 4.8663282), List(5490, 4.8567095), List(928, 4.7913594), List(2239, 4.761064), List(28, 4.7050796), List(1949, 4.701012), List(1267, 4.6946077), List(6650, 4.6499), List(7008, 4.605741), List(3200, 4.6008387)) 85 List(List(51931, 5.422657), List(112804, 5.334261), List(89904, 5.1279993), List(2135, 5.074174), List(47423, 5.0500717), List(1282, 4.960841), List(25825, 4.947359), List(53, 4.9196067), List(2583, 4.899414), List(1046, 4.8935494)) 65 List(List(177593, 4.9033103), List(3200, 4.6735377), List(179135, 4.6531224), List(171495, 4.6531224), List(26073, 4.6531224), List(72171, 4.6531224), List(7071, 4.6531224), List(84273, 4.6531224), List(117531, 4.6531224), List(184245, 4.6531224)) 458 List(List(3200, 5.7471757), List(87234, 5.6294656), List(5075, 5.614453), List(7743, 5.293073), List(928, 5.230948), List(25906, 5.2069235), List(77846, 5.2069235), List(93008, 5.2069235), List(1066, 5.2015905), List(1046, 5.189669)) 481 List(List(969, 4.428582), List(213, 4.414027), List(3099, 4.3728013), List(4495, 4.283353), List(6201, 4.283353), List(215,

In [55]:
userRecs.createOrReplaceTempView("userRec")
temp = spark.sql("SELECT userId, recommendations FROM userRec where userId = 575")
display(temp)

userId recommendations 575 List(List(8477, 5.420325), List(183897, 5.228679), List(86781, 5.1137958), List(177593, 5.080568), List(148881, 5.067505), List(2303, 5.021811), List(7025, 5.002827), List(188751, 4.9677463), List(176371, 4.9407716), List(58301, 4.9085007))

In [56]:
userRecs.createOrReplaceTempView("userRec")
temp = spark.sql("SELECT userId, recommendations FROM userRec where userId = 232")

display(temp)

userId recommendations 232 List(List(3200, 4.697607), List(5075, 4.654945), List(171495, 4.548487), List(84273, 4.548487), List(26073, 4.548487), List(72171, 4.548487), List(117531, 4.548487), List(179135, 4.548487), List(7071, 4.548487), List(184245, 4.548487))

## Find the similar moives for moive with id: 463, 471
You can find the similar moives based on the ALS results

In [58]:
factors=best_model.itemFactors
display(factors)

id features 10 List(0.29814208, 0.36530912, 0.1235385, 1.3191936, 0.41615722, 0.049333323, 0.7169036, 0.66238296, 0.58697015, 0.4086679, 0.30274802, 0.24906948) 20 List(0.5149267, 0.4681858, 0.66447353, 0.024519358, 0.2475168, 0.0, 0.49447954, 0.7850115, 0.3085634, 0.38451654, 1.2286403, 0.0) 30 List(1.9744086, 0.0, 0.6769193, 0.0, 0.0, 0.1134667, 0.44481862, 0.5916831, 0.23605524, 0.0, 0.0, 1.018136) 40 List(0.14861526, 0.018709095, 0.8189067, 1.1998013, 0.34267634, 0.54863214, 0.00358345, 1.2906394, 0.009430693, 0.5723096, 0.08932574, 0.5129575) 50 List(0.7055091, 0.38433766, 1.0771112, 0.97160816, 0.47269776, 0.18892238, 0.4336216, 0.9773563, 0.8559772, 0.4103268, 0.29588607, 0.3446865) 60 List(0.37999052, 0.49173614, 0.18709163, 0.642562, 0.42471606, 0.49961355, 0.39475325, 0.832857, 0.23314984, 0.757091, 0.8381465, 0.29273462) 70 List(0.26945016, 0.40862557, 0.08552736, 0.90924174, 0.32633874, 0.34024343, 0.56268543, 0.89316005, 0.98951846, 0.19658491, 1.025683, 0.0) 80 List(0.56168956, 0.16450606, 0.9886159, 0.5288544, 0.33633593, 0.47990605, 0.642177, 0.5978784, 0.67271215, 0.4380014, 0.3626409, 0.6309376) 100 List(0.22152518, 0.7228624, 0.37474236, 0.38841593, 0.35063666, 0.5904562, 0.51418054, 0.2201264, 0.1676978, 0.5865636, 0.22820547, 0.5933339) 110 List(0.2862576, 0.3276106, 0.9418058, 1.0464691, 0.26508057, 0.0, 0.7774756, 0.70251465, 0.54621005, 0.975931, 0.35623717, 0.27825004) 140 List(0.49053463, 0.5157049, 0.7233471, 0.8376465, 0.61697453, 0.16461505, 0.2220656, 0.3672975, 0.069199905, 0.84926933, 0.30462444, 0.4011406) 150 List(0.64193565, 0.4225855, 0.59961635, 0.8637335, 0.6432525, 0.0076820394, 0.3582702, 0.6367129, 0.41018, 1.1151935, 0.27877942, 0.3240934) 160 List(0.0, 0.45610487, 0.0, 1.1205014, 0.352416, 0.33682546, 0.26612738, 0.0, 0.23853758, 0.8080178, 0.59022844, 0.34047627) 170 List(0.59220535, 0.18690795, 0.42776486, 0.9161985, 0.3624892, 0.21730489, 0.89274627, 0.23868054, 0.4325979, 0.33318296, 0.63302886, 0.47318026) 180 List(0.9059805, 0.2969084, 0.44174957, 0.8728348, 0.0, 0.019890733, 0.5654472, 0.5303909, 1.1918094, 0.15798928, 0.45465285, 0.1241589) 190 List(0.7475659, 0.0, 0.8180856, 0.0, 0.0, 0.2442989, 1.5314337, 0.21395911, 0.7362614, 0.112741455, 0.0, 0.8649518) 210 List(0.24808525, 0.34551656, 0.13724913, 0.9251853, 0.41234186, 0.2621423, 0.27992308, 0.21307632, 0.39834332, 0.7211099, 0.43720022, 0.76065975) 220 List(0.12592572, 0.20035043, 0.10909862, 0.36535951, 0.37239966, 0.31599182, 0.40776235, 0.38996282, 0.64702463, 0.35588282, 0.60099703, 0.38597527) 230 List(0.32146403, 0.16832323, 0.3285144, 0.9446352, 0.4935349, 0.23930934, 0.39465392, 0.38852057, 0.33117634, 0.91090536, 0.5702897, 0.83130145) 240 List(0.0, 0.58548886, 0.0, 0.4936883, 0.09621142, 0.16680294, 0.1172434, 0.21558233, 0.22785525, 0.74979115, 0.54591537, 0.4175471) 250 List(0.10813843, 0.4997644, 0.15358755, 0.8261851, 0.37631318, 0.3922518, 0.32583395, 0.20881349, 0.36763203, 0.8159327, 0.53202057, 0.6144314) 260 List(0.94145566, 0.49075043, 0.89096975, 1.0585641, 0.69499546, 0.0, 0.24769421, 0.40420774, 0.94981015, 0.9346653, 0.14595078, 0.15098603) 270 List(0.1355042, 0.6671995, 0.0, 0.6101214, 1.0409932, 0.6431173, 0.2120691, 0.3323219, 0.0, 0.8900648, 0.37059236, 0.6116288) 280 List(0.33611718, 0.41861677, 0.49452916, 1.1162012, 0.46417528, 0.5457672, 0.6712433, 0.5267132, 0.52403104, 0.30823207, 0.53167665, 1.0601428) 290 List(1.1357379, 0.0, 0.55198747, 0.7153501, 0.4332575, 0.7171933, 0.22971521, 0.32874015, 0.9568419, 0.87278616, 0.0, 0.8078209) 300 List(0.2117848, 0.22598909, 1.1463108, 0.61499435, 0.44573233, 0.0, 0.6059962, 0.57744783, 0.77689195, 0.59078056, 0.012146172, 0.5566678) 310 List(0.055783544, 0.38265032, 0.104537554, 0.69368285, 0.25424075, 0.2622348, 0.24453811, 0.1485568, 0.21913923, 0.6148617, 0.41734463, 0.5410123) 320 List(0.54649615, 0.34275863, 0.14031966, 0.018571462, 0.2411236, 0.40597543, 0.2673254, 0.70815563, 0.5808374, 0.22631066, 0.17107794, 0.31147087) 330 List(0.08376

In [59]:
factors.filter(col('id') == 463).show()

+---+--------+
 id|features|
+---+--------+
+---+--------+

In [60]:
factors.filter(col('id') == 471).show()

+---+--------------------+
 id| features|
+---+--------------------+
471|[0.3588241, 0.456...|
+---+--------------------+

In [61]:

import numpy as np


In [62]:
def similar_movies(factors, movieId):

  try: 
    target_id_feature = np.array(factors.filter(col('id') == movieId).select("features").collect())[0]
  except:
    return 'There is no movie with id ' + str(movieId)

  similarities = []
  for feature in np.array(factors.select("features").collect()):
    feature = feature.reshape(12)
    target_id_feature = target_id_feature.reshape(12)
    similarity = np.dot(target_id_feature,feature)/(np.linalg.norm(target_id_feature) * np.linalg.norm(feature))
    similarities.append(similarity)
  
  index = np.array(factors.select("id").collect())
  np_s = np.array(similarities).reshape(8948,1)
  combined_data = np.zeros([8948,2])
  combined_data.shape
  combined_data[:,0]=index[:,0]
  combined_data[:,1]=np_s[:,0]
  pdf = pd.DataFrame(combined_data)
  pdf.columns=['movie index','similarity']
  
  # top 11 similar movies contain the movie itself with similarity = 1, so I need to remove it. 
  top10_movie = pdf.nlargest(11, 'similarity')
  
  return top10_movie

In [63]:
similar_movies(factors, 471)


movie index 
 similarity 
 
 
 
 
 934 
 471.0 
 1.000000 
 
 
 994 
 1271.0 
 0.939895 
 
 
 8133 
 639.0 
 0.937794 
 
 
 3301 
 44633.0 
 0.936442 
 
 
 8693 
 58309.0 
 0.936442 
 
 
 5932 
 26366.0 
 0.936442 
 
 
 7502 
 4308.0 
 0.935694 
 
 
 946 
 661.0 
 0.930107 
 
 
 6388 
 1307.0 
 0.927974 
 
 
 6960 
 72737.0 
 0.925677 
 
 
 6294 
 17.0 
 0.921007

In [64]:
similar_movies(factors, 463)

Out[155]: 'There is no movie with id 463'

****Project Outlines***

When we are doing online shopping or visiting video websites, like Amazon and Hulu, we always see a list named “You may also like”. I am very curious about how the online shopping or video platform come up with this list. And I want to use the data analysis technique to recommend products to different users and find out the similar products for the specific products.
1. At first, I download the movie data and create four data frames, movies_df,  ratings_df, links_df and tags-df. And I use some basic methods to give a brief glance of the data.
2. After that, I used spark SQL to do OLAP of the data. I found out the total number of users is 610 and the total number of movies is 9742. After some SQL operation, we can see there are around 20 movies which have no rating. And I also used spark SQL to build a table with lists of all movies for each category.
3. Then we used park ALS to train the movie data and hopefully we can use this model to predict all the ratings for all users and find out the similar movies.  First, we import the evaluator, ALS and the tuning packages. Then we create test and train set. After that, we turn the model using paramGridBuilder. Then we fit the model by cross validation. Then we extract best model, generate predictions and evaluate using RMSE. After model testing, we apply the model to see the performance. 
4. Finally, we can use the model to recommend the movies to specific users. Most importantly, using the dot production from the features matrix, which we extracted from the ALS models, we can find out the similarities among diferent movie items. This feature is very useful when customers want to explode the similar products based on their existing favorite products.

This ALS model can be useful to recommend movie to a current user. However, it has a coldstart issue and it is not applicable for a new user with no former information.